lib/exabgp/bgp/message/update/attribute/bgpls/link/sradj.py

```diff
@@ -29,20 +29,18 @@
 @LINKSTATE.register()
 class SrAdjacency(object):
 	TLV = 1099
+	_sr_adj_sids = []
```

Comment:
is there a way to maintain this state on the instance instead - and maybe reseting the instance - ? (or worse case outside of the class itself) having it as part of the class object reads a bit weird.

Reply:
As an update message is parsed containing multiple TLVs of the same type, multiple instances of the corresponding class will be created. As each instance is created, via the `unpack` constructor, state is appended to the class variable `_sr_adj_sids`. The final instance of the class then has state from all previous instances of that class. Only the last instance of the class should be output. I just amended the PR as I forgot to do this (and my output was hiding the duplicates). The `reset` method clears the class variable for the next update message that needs to use the class.

Maintaining state within the instance will not work with my implementation, as each instance needs to know of previous instance's state. State could be tracked elsewhere, but this would require a fair bit of a refactor of the linkstate.py module as it's `json` and `__str__` methods are expecting each TLV class instance to return a json string with a unique key. I felt this was the least intrusive fix but happy to refactor if you have better ideas

Reply:
@thomas-mangin @ruissalo I was thinking about this some more. for the classes that potentially represent more than one TLV, singletons could be an option. this would remove the need for this line https://github.com/Exa-Networks/exabgp/pull/964/commits/b539901955f9dc33a910c4fd4a798505b6551bf8#diff-a590d43e4a788bd5d3de7cdc2bafb9caR63 (which feels a little hacky) as long as ```ls_attrs``` was initialized as a set instead of a list. any thoughts?

A few singleton recipes here - https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Singleton.html

Reply:
Or another option would be to use a default list argument within the ```__init__``` method. The thing you are told to never do ;)

something like:
```
    def __init__(self, sr_adj_sid, sr_adj_sids=[]):
        self.sr_adj_sids = sr_adj_sids.append(sr_adj_sid)
```